File: server/src/utils/logger.ts
/**
 * @fileoverview Logging utility with timestamps and timing support.
 * Provides structured, colorful console output for tracking analysis stages.
 * Optionally writes logs to a timestamped file when WRITE_LOG_TO_FILE is set.
 */

import * as fs from 'fs'
import * as path from 'path'

// ============================================================================
// Types
// ============================================================================

/** Log level for categorizing messages */
type LogLevel = 'info' | 'success' | 'warn' | 'error' | 'debug' | 'timing'

/** Timer storage for tracking operation durations */
const timers: Map<string, number> = new Map()

// ============================================================================
// File Logging Setup
// ============================================================================

/** Whether to write logs to file */
const writeToFile = process.env.WRITE_LOG_TO_FILE === 'true'

/** Log file path (created once at startup) */
let logFilePath: string | null = null

/** Log file write stream */
let logFileStream: fs.WriteStream | null = null

/**
 * Initialize file logging if enabled.
 */
function initFileLogging(): void {
  if (!writeToFile || logFileStream) return
  
  // Create logs directory if it doesn't exist
  const logsDir = path.resolve(process.cwd(), 'logs')
  if (!fs.existsSync(logsDir)) {
    fs.mkdirSync(logsDir, { recursive: true })
  }
  
  // Create timestamped log file
  const now = new Date()
  const timestamp = now.toISOString()
    .replace(/[:.]/g, '-')
    .replace('T', '_')
    .slice(0, 19)
  logFilePath = path.join(logsDir, `meddlingkids_${timestamp}.log`)
  
  logFileStream = fs.createWriteStream(logFilePath, { flags: 'a' })
  
  // Write header
  const header = `
================================================================================
  Meddling Kids Log - Started ${now.toISOString()}
================================================================================
`
  logFileStream.write(header)
  
  // Log to console that file logging is enabled
  console.log(`\x1b[36mℹ [Logger] Writing logs to: ${logFilePath}\x1b[0m`)
}

/**
 * Write a line to the log file (without ANSI colors).
 */
function writeToLogFile(line: string): void {
  if (!logFileStream) return
  
  // Strip ANSI escape codes for file output
  // eslint-disable-next-line no-control-regex
  const cleanLine = line.replace(/\x1b\[[0-9;]*m/g, '')
  logFileStream.write(cleanLine + '\n')
}

// Initialize file logging on module load
initFileLogging()

// ============================================================================
// ANSI Colors
// ============================================================================

const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  dim: '\x1b[2m',
  cyan: '\x1b[36m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  red: '\x1b[31m',
  magenta: '\x1b[35m',
  blue: '\x1b[34m',
  gray: '\x1b[90m',
}

// ============================================================================
// Formatting Helpers
// ============================================================================

/**
 * Get current timestamp in HH:MM:SS.mmm format.
 */
function getTimestamp(): string {
  const now = new Date()
  const hours = now.getHours().toString().padStart(2, '0')
  const minutes = now.getMinutes().toString().padStart(2, '0')
  const seconds = now.getSeconds().toString().padStart(2, '0')
  const ms = now.getMilliseconds().toString().padStart(3, '0')
  return `${hours}:${minutes}:${seconds}.${ms}`
}

/**
 * Format duration in milliseconds to human-readable string.
 */
function formatDuration(ms: number): string {
  if (ms < 1000) {
    return `${ms}ms`
  }
  if (ms < 60000) {
    return `${(ms / 1000).toFixed(2)}s`
  }
  const minutes = Math.floor(ms / 60000)
  const seconds = ((ms % 60000) / 1000).toFixed(1)
  return `${minutes}m ${seconds}s`
}

/**
 * Get color code for log level.
 */
function getLevelColor(level: LogLevel): string {
  switch (level) {
    case 'info': return colors.cyan
    case 'success': return colors.green
    case 'warn': return colors.yellow
    case 'error': return colors.red
    case 'debug': return colors.gray
    case 'timing': return colors.magenta
  }
}

/**
 * Get level prefix symbol.
 */
function getLevelSymbol(level: LogLevel): string {
  switch (level) {
    case 'info': return 'ℹ'
    case 'success': return '✓'
    case 'warn': return '⚠'
    case 'error': return '✗'
    case 'debug': return '•'
    case 'timing': return '⏱'
  }
}

// ============================================================================
// Core Logger
// ============================================================================

/**
 * Logger class for structured console output.
 */
class Logger {
  private context: string

  constructor(context: string = 'Server') {
    this.context = context
  }

  /**
   * Create a child logger with a new context.
   */
  child(context: string): Logger {
    return new Logger(context)
  }

  /**
   * Internal log method.
   */
  private log(level: LogLevel, message: string, data?: Record<string, unknown>): void {
    const timestamp = getTimestamp()
    const color = getLevelColor(level)
    const symbol = getLevelSymbol(level)
    
    const prefix = `${colors.gray}[${timestamp}]${colors.reset} ${color}${symbol}${colors.reset} ${colors.bright}[${this.context}]${colors.reset}`
    
    let logLine: string
    if (data && Object.keys(data).length > 0) {
      const dataStr = Object.entries(data)
        .map(([k, v]) => `${colors.dim}${k}=${colors.reset}${formatValue(v)}`)
        .join(' ')
      logLine = `${prefix} ${message} ${dataStr}`
    } else {
      logLine = `${prefix} ${message}`
    }
    
    console.log(logLine)
    writeToLogFile(logLine)
  }

  /** Log info message */
  info(message: string, data?: Record<string, unknown>): void {
    this.log('info', message, data)
  }

  /** Log success message */
  success(message: string, data?: Record<string, unknown>): void {
    this.log('success', message, data)
  }

  /** Log warning message */
  warn(message: string, data?: Record<string, unknown>): void {
    this.log('warn', message, data)
  }

  /** Log error message */
  error(message: string, data?: Record<string, unknown>): void {
    this.log('error', message, data)
  }

  /** Log debug message */
  debug(message: string, data?: Record<string, unknown>): void {
    this.log('debug', message, data)
  }

  /**
   * Start a timer for an operation.
   * @param label - Unique label for the timer
   */
  startTimer(label: string): void {
    const key = `${this.context}:${label}`
    timers.set(key, Date.now())
    this.log('timing', `Starting: ${label}`)
  }

  /**
   * End a timer and log the duration.
   * @param label - The timer label (must match startTimer)
   * @param message - Optional completion message
   * @returns Duration in milliseconds
   */
  endTimer(label: string, message?: string): number {
    const key = `${this.context}:${label}`
    const start = timers.get(key)
    
    if (!start) {
      this.warn(`Timer "${label}" was not started`)
      return 0
    }
    
    const duration = Date.now() - start
    timers.delete(key)
    
    const durationStr = `${colors.magenta}${formatDuration(duration)}${colors.reset}`
    const displayMessage = message || `Completed: ${label}`
    this.log('timing', `${displayMessage} ${colors.dim}took${colors.reset} ${durationStr}`)
    
    return duration
  }

  /**
   * Log a section header for visual separation.
   */
  section(title: string): void {
    const line = '─'.repeat(60)
    const output = [
      '',
      `${colors.blue}${line}${colors.reset}`,
      `${colors.blue}${colors.bright}  ${title}${colors.reset}`,
      `${colors.blue}${line}${colors.reset}`,
      ''
    ]
    output.forEach(l => {
      console.log(l)
      writeToLogFile(l)
    })
  }

  /**
   * Log a subsection header.
   */
  subsection(title: string): void {
    const output = `\n${colors.cyan}  ▸ ${title}${colors.reset}`
    console.log(output)
    writeToLogFile(output)
  }
}

/**
 * Format a value for display.
 */
function formatValue(value: unknown): string {
  if (value === null || value === undefined) {
    return `${colors.dim}null${colors.reset}`
  }
  if (typeof value === 'number') {
    return `${colors.yellow}${value}${colors.reset}`
  }
  if (typeof value === 'boolean') {
    return value ? `${colors.green}true${colors.reset}` : `${colors.red}false${colors.reset}`
  }
  if (typeof value === 'string') {
    // Truncate long strings
    const display = value.length > 50 ? value.substring(0, 47) + '...' : value
    return `${colors.green}"${display}"${colors.reset}`
  }
  if (Array.isArray(value)) {
    return `${colors.cyan}[${value.length} items]${colors.reset}`
  }
  if (typeof value === 'object') {
    const keys = Object.keys(value).length
    return `${colors.cyan}{${keys} keys}${colors.reset}`
  }
  return String(value)
}

// ============================================================================
// Exports
// ============================================================================

/** Main logger instance */
export const logger = new Logger('Server')

/** Create a logger for a specific module */
export function createLogger(context: string): Logger {
  return new Logger(context)
}

File: server/src/utils/logger.py
"""
Logging utility with timestamps and timing support.
Provides structured, colourful console output for tracking analysis stages.
Optionally writes logs to a timestamped file when WRITE_TO_FILE is set.

All mutable per-session state (timers, log buffer, log-file handle)
is stored in ``contextvars.ContextVar`` so that concurrent async
analysis tasks do not interfere with each other.
"""

from __future__ import annotations

import contextvars
import io
import os
import pathlib
import re
import sys
import time
from datetime import UTC, datetime

# ============================================================================
# Per-session state (isolated via contextvars)
# ============================================================================

_timers_var: contextvars.ContextVar[dict[str, tuple[float, str]]] = contextvars.ContextVar("_timers_var")
_log_buffer_var: contextvars.ContextVar[list[str]] = contextvars.ContextVar("_log_buffer_var")
_log_file_stream_var: contextvars.ContextVar[io.TextIOWrapper | None] = contextvars.ContextVar("_log_file_stream_var", default=None)
_log_file_path_var: contextvars.ContextVar[str | None] = contextvars.ContextVar("_log_file_path_var", default=None)


def _get_timers() -> dict[str, tuple[float, str]]:
    """Return the per-context timer dict, creating it on first access."""
    try:
        return _timers_var.get()
    except LookupError:
        timers: dict[str, tuple[float, str]] = {}
        _timers_var.set(timers)
        return timers


def _get_log_buffer() -> list[str]:
    """Return the per-context log buffer, creating it on first access."""
    try:
        return _log_buffer_var.get()
    except LookupError:
        buf: list[str] = []
        _log_buffer_var.set(buf)
        return buf


def get_log_buffer() -> list[str]:
    """Return a copy of the accumulated log lines (ANSI-stripped)."""
    return list(_get_log_buffer())


def clear_log_buffer() -> None:
    """Clear the in-memory log buffer and timers for the next analysis run."""
    _get_log_buffer().clear()
    _get_timers().clear()


# ============================================================================
# File Logging
# ============================================================================

_write_to_file = os.environ.get("WRITE_TO_FILE", "").lower() == "true"


def start_log_file(domain: str) -> None:
    """Start a new log file for a specific analysis."""
    if not _write_to_file:
        return

    end_log_file()

    logs_dir = pathlib.Path.cwd() / ".logs"
    logs_dir.mkdir(parents=True, exist_ok=True)

    safe_domain = domain.removeprefix("www.")
    safe_domain = "".join(c if c.isalnum() or c in ".-" else "_" for c in safe_domain)[:50]

    now = datetime.now(UTC)
    timestamp = now.strftime("%Y-%m-%d_%H-%M-%S")
    log_file_path = str(logs_dir / f"{safe_domain}_{timestamp}.log")

    try:
        stream = open(log_file_path, "a", encoding="utf-8")  # noqa: SIM115
    except OSError as exc:
        print(f"\033[31m✗ [Logger] Failed to open log file: {exc}\033[0m")
        return

    _log_file_stream_var.set(stream)
    _log_file_path_var.set(log_file_path)

    header = f"\n{'=' * 80}\n  Analysis Log - {domain}\n  Started: {now.isoformat()}\n{'=' * 80}\n"
    stream.write(header)
    print(f"\033[36mℹ [Logger] Writing logs to: {log_file_path}\033[0m")


def end_log_file() -> None:
    """Flush and close the current log file."""
    stream = _log_file_stream_var.get(None)
    if stream is not None:
        try:
            stream.flush()
            stream.close()
        except Exception:
            print("\033[33m⚠ [Logger] Failed to flush/close log file stream\033[0m")
        _log_file_stream_var.set(None)
        _log_file_path_var.set(None)


def save_report_file(
    domain: str,
    report_text: str,
) -> str | None:
    """Save the final structured report as a text file.

    Only writes when ``WRITE_TO_FILE`` is enabled.

    Args:
        domain: The analysed domain (used in filename).
        report_text: Rendered report content.

    Returns:
        The file path written, or ``None`` if skipped.
    """
    if not _write_to_file:
        return None

    reports_dir = pathlib.Path.cwd() / ".reports"
    reports_dir.mkdir(parents=True, exist_ok=True)

    safe_domain = domain.removeprefix("www.")
    safe_domain = "".join(c if c.isalnum() or c in ".-" else "_" for c in safe_domain)[:50]

    now = datetime.now(UTC)
    timestamp = now.strftime("%Y-%m-%d_%H-%M-%S")
    path = reports_dir / f"{safe_domain}_{timestamp}.txt"

    try:
        path.write_text(report_text, encoding="utf-8")
        print(f"\033[36m\u2139 [Logger] Report saved to: {path}\033[0m")
        return str(path)
    except Exception as err:
        print(f"\033[31m\u2717 [Logger] Failed to save report: {err}\033[0m")
        return None


def _write_to_log_file(line: str) -> None:
    """Write a line to the log file (without ANSI colours)."""
    stream = _log_file_stream_var.get(None)
    if stream is None:
        return

    clean = re.sub(r"\033\[[0-9;]*m", "", line)
    stream.write(clean + "\n")
    stream.flush()


# ============================================================================
# ANSI Colours
# ============================================================================

_colours = {
    "reset": "\033[0m",
    "bright": "\033[1m",
    "dim": "\033[2m",
    "cyan": "\033[36m",
    "green": "\033[32m",
    "yellow": "\033[33m",
    "red": "\033[31m",
    "magenta": "\033[35m",
    "blue": "\033[34m",
    "gray": "\033[90m",
}

_level_colour = {
    "info": _colours["cyan"],
    "success": _colours["green"],
    "warn": _colours["yellow"],
    "error": _colours["red"],
    "debug": _colours["gray"],
    "timing": _colours["magenta"],
}

_level_symbol = {
    "info": "ℹ",
    "success": "✓",
    "warn": "⚠",
    "error": "✗",
    "debug": "•",
    "timing": "⏱",
}


def _get_timestamp() -> str:
    """Return the current UTC time as HH:MM:SS.mmm."""
    now = datetime.now(UTC)
    return now.strftime("%H:%M:%S.") + f"{now.microsecond // 1000:03d}"


def _format_duration(ms: float) -> str:
    """Format a duration in milliseconds for display."""
    if ms < 1000:
        return f"{int(ms)}ms"
    if ms < 60000:
        return f"{ms / 1000:.2f}s"
    minutes = int(ms // 60000)
    seconds = (ms % 60000) / 1000
    return f"{minutes}m {seconds:.1f}s"


def _format_value(value: object) -> str:
    """Return an ANSI-coloured representation of *value*."""
    c = _colours
    if value is None:
        return f"{c['dim']}None{c['reset']}"
    if isinstance(value, bool):
        return f"{c['green']}True{c['reset']}" if value else f"{c['red']}False{c['reset']}"
    if isinstance(value, (int, float)):
        return f"{c['yellow']}{value}{c['reset']}"
    if isinstance(value, str):
        display = value[:497] + "..." if len(value) > 500 else value
        return f'{c["green"]}"{display}"{c["reset"]}'
    if isinstance(value, list):
        return f"{c['cyan']}[{len(value)} items]{c['reset']}"
    if isinstance(value, dict):
        return f"{c['cyan']}{{{len(value)} keys}}{c['reset']}"
    return str(value)


# ============================================================================
# Logger Class
# ============================================================================


class Logger:
    """Structured logger with context prefix and timing support."""

    def __init__(self, context: str = "Server") -> None:
        """Create a logger that prefixes messages with *context*."""
        self._context = context

    def _log(self, level: str, message: str, data: dict[str, object] | None = None) -> None:
        """Format and emit a log line at the given level."""
        ts = _get_timestamp()
        colour = _level_colour.get(level, _colours["cyan"])
        symbol = _level_symbol.get(level, "ℹ")
        c = _colours

        prefix = f"{c['gray']}[{ts}]{c['reset']} {colour}{symbol}{c['reset']} {c['bright']}[{self._context}]{c['reset']}"

        if data:
            data_str = " ".join(f"{c['dim']}{k}={c['reset']}{_format_value(v)}" for k, v in data.items())
            log_line = f"{prefix} {message} {data_str}"
        else:
            log_line = f"{prefix} {message}"

        print(log_line, file=sys.stderr)
        _write_to_log_file(log_line)
        # Keep an ANSI-stripped copy for the client debug tab.
        _get_log_buffer().append(re.sub(r"\033\[[0-9;]*m", "", log_line))

    def info(self, message: str, data: dict[str, object] | None = None) -> None:
        """Log an informational message."""
        self._log("info", message, data)

    def success(self, message: str, data: dict[str, object] | None = None) -> None:
        """Log a success message."""
        self._log("success", message, data)

    def warn(self, message: str, data: dict[str, object] | None = None) -> None:
        """Log a warning message."""
        self._log("warn", message, data)

    def error(self, message: str, data: dict[str, object] | None = None) -> None:
        """Log an error message."""
        self._log("error", message, data)

    def debug(self, message: str, data: dict[str, object] | None = None) -> None:
        """Log a debug-level message."""
        self._log("debug", message, data)

    def start_timer(self, label: str) -> None:
        """Start a named timer for performance measurement."""
        key = f"{self._context}:{label}"
        _get_timers()[key] = (time.monotonic() * 1000, _get_timestamp())
        self._log("timing", f"Starting: {label}")

    def end_timer(self, label: str, message: str | None = None) -> float:
        """Stop a named timer and log the elapsed time."""
        key = f"{self._context}:{label}"
        entry = _get_timers().pop(key, None)
        if entry is None:
            self.warn(f'Timer "{label}" was not started')
            return 0.0

        start_ms, start_ts = entry
        duration = time.monotonic() * 1000 - start_ms
        c = _colours
        duration_str = f"{c['magenta']}{_format_duration(duration)}{c['reset']}"
        display_message = message or f"Completed: {label}"
        self._log(
            "timing",
            f"{display_message} {c['dim']}took{c['reset']} {duration_str} {c['dim']}(started {start_ts}){c['reset']}",
        )
        return duration

    def section(self, title: str) -> None:
        """Print a prominent section divider with *title*."""
        c = _colours
        line = "─" * 60
        lines = [
            "",
            f"{c['blue']}{line}{c['reset']}",
            f"{c['blue']}{c['bright']}  {title}{c['reset']}",
            f"{c['blue']}{line}{c['reset']}",
            "",
        ]
        for ln in lines:
            print(ln, file=sys.stderr)
            _write_to_log_file(ln)
            _get_log_buffer().append(re.sub(r"\033\[[0-9;]*m", "", ln))

    def subsection(self, title: str) -> None:
        """Print a smaller sub-section header."""
        c = _colours
        output = f"\n{c['cyan']}  ▸ {title}{c['reset']}"
        print(output, file=sys.stderr)
        _write_to_log_file(output)
        _get_log_buffer().append(re.sub(r"\033\[[0-9;]*m", "", output))


def create_logger(context: str) -> Logger:
    """Create a logger for a specific module."""
    return Logger(context)
